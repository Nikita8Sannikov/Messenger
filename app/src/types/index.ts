// В данном файле хранятся общие абстрактные типы, применимые на всем уровне клиента

export type ValueOf<T> = T[keyof T];

export type TimeoutType = ReturnType<typeof setTimeout>;
