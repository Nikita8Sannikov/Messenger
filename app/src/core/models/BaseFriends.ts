import { ApiRoutes } from "common-types";

import { type IFriend } from "@custom-types/friends.types";

export interface IGetAll<T> { 
    route: ApiRoutes;
    page?: number;
    data?: { ids: string[] };
    setLoading: (isLoading: boolean) => void; 
    successCb: (data: { success: boolean; } & T) => void; 
};

// Контракт модели "Управление друзьями"
export interface BaseFriends {
    setCount(count: number): void;
    add(user: IFriend): void;
    remove(userId: string): void;
    find(userId: string): IFriend | undefined;
};