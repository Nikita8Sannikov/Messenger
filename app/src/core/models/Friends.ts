import { type FriendsTab } from "common-types";

import type FriendsController from "@core/controllers/FriendsController";
import { type IUser } from "@custom-types/models.types";

// Контракт модели "Сервис друзей"
export interface Friends {
    friendsController: FriendsController;
    
    checkOnlineUsers(users: IUser[]): void;
    removeOnlineUser(userId: string): void;
    getFriends(type: FriendsTab): void;
    loadMore(type: FriendsTab, resolve: () => void): void;
    search(type: FriendsTab, value: string): void;

    followFriend(friendId: string): void;
    writeMessage(friendId: string): void;
    addFriend(friendId: string): void;
    accept(friendId: string): void;
    leftInFollowers(friendId: string): void;
    unfollow(friendId: string): void;
    deleteFriend(friendId: string): void;
    blockFriend(friendId: string): void;
    unblock(friendId: string): void;
};