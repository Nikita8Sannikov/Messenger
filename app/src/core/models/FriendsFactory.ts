import { type BaseFriends } from "@core/models/BaseFriends";

// Контракт модели "Фабрика друзей"
export interface FriendsFactory {
    createFriendsManager(): BaseFriends;
    createOutgoingRequestsManager(): BaseFriends;
    createFollowersManager(): BaseFriends;
    createBlockedUsersManager(): BaseFriends;
    createOnlineFriendsManager(): BaseFriends;
    createCommonFriendsManager(): BaseFriends;
};