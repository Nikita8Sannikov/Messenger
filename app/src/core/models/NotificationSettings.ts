import { type INotificationSettings } from "@custom-types/models.types";

// Контракт модели "Настройки уведомлений пользователя"
export interface NotificationSettings {
	settings: INotificationSettings;
};