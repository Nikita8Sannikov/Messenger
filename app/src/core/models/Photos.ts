import { type IPhoto } from "@custom-types/models.types";

// Контракт модели "Фотографии"
export interface Photos {
	photos: IPhoto[];
	count: number;

	getAllPhotos: (resolve?: () => void) => void;
	onClickPhoto: (photoId: string) => void;
	addPhotoFromAvatar: (photo: IPhoto) => void;
	addPhotos: (data: FormData) => void;
	deletePhoto: (data: { photoId: string; imageUrl: string; isAvatar: boolean; }, fromProfile?: boolean) => void;
};