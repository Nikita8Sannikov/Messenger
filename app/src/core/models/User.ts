import { type NotificationSettings } from "@core/models/NotificationSettings";
import { type Photos } from "@core/models/Photos";
import { type UserDetails } from "@core/models/UserDetails";
import type { IApiUser, IApiUserDetails } from "@custom-types/api.types";

// Контракт модели "Пользователь"
export interface User {
	photos: Photos;
	details: UserDetails;
	settings?: NotificationSettings;

	id: string;
	firstName: string;
	thirdName: string;
	phone: string;
	email: string;
	fullName: string;
	avatarUrl: string;
	avatarCreateDate: string;

	updateUser: ({ user, userDetails }: { user: IApiUser; userDetails: IApiUserDetails; }) => void;
	changeAvatar: (updatedAvatar?: { newAvatarUrl: string; avatarCreationDate: string; }) => void;
};
