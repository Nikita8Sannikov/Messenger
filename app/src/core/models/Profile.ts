import EventEmitter from "eventemitter3";
import { Dispatch, SetStateAction } from "react";
import dayjs from "dayjs";

import { User } from "@core/models/User";
import { IFormValues } from "@pages/Edit";
import { IPhoto } from "@custom-types/models.types";

// Контракт модели "Профиль пользователя"
export interface Profile extends EventEmitter {
    user: User;

    onClickAvatar: () => void;
    onDeleteAvatar: (setLoading?: Dispatch<SetStateAction<boolean>>) => void;
    onSetAvatar: (newAvatar: { id: string; newAvatarUrl: string; newPhotoUrl: string; }) => void;
    getAllPhotos: (setLoading: Dispatch<SetStateAction<boolean>>) => void;
    updatePhotosCount: (count: number) => void;
    onClickPhoto: (photos: IPhoto[], index: number) => void;
    addPhotos: (data: Object, setLoading: Dispatch<SetStateAction<boolean>>) => void;
    deletePhoto: (data: Object, photos: IPhoto[], path: string) => void;
    getFriends: (setLoading: Dispatch<SetStateAction<boolean>>) => void;
    editInfo: ( result: IFormValues) => void;
}