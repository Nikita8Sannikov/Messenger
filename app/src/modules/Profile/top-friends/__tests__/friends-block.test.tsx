import { render, screen } from "@testing-library/react";

import FriendsBlock from "@modules/profile/top-friends/friends-block";
import { IUser } from "@custom-types/models.types";

const mockUsers: IUser[] = [
	{
		id: "1",
		firstName: "John",
		secondName: "Doe",
		thirdName: "Smith",
		email: "john@example.com",
		phone: "1234567890",
		avatarUrl: "avatar1.jpg",
		avatarCreateDate: "2024-01-01",
		fullName: "John Doe Smith",
	},
	{
		id: "2",
		firstName: "Jane",
		secondName: "Smith",
		thirdName: "Doe",
		email: "jane@example.com",
		phone: "0987654321",
		avatarUrl: "avatar2.jpg",
		avatarCreateDate: "2024-01-02",
		fullName: "Jane Smith Doe",
	},
];

describe("FriendsBlock", () => {
	it("renders with title and count", () => {
		const { container } = render(
			<FriendsBlock
				state={{
					title: "Friends",
					count: 2,
					users: mockUsers,
				}}
				onClickBlock={() => {}}
			>
				<div>No friends</div>
			</FriendsBlock>,
		);
		expect(container).toMatchSnapshot();
		expect(screen.getByText("Friends")).toBeInTheDocument();
		expect(screen.getByText("2")).toBeInTheDocument();
	});

	it("renders loading state", () => {
		const { container } = render(
			<FriendsBlock
				state={{
					title: "Friends",
					count: 2,
					users: mockUsers,
				}}
				isLoading={true}
				onClickBlock={() => {}}
			>
				<div>No friends</div>
			</FriendsBlock>,
		);
		expect(container).toMatchSnapshot();
		expect(screen.getByRole("progressbar")).toBeInTheDocument();
	});

	it("renders users list when users are provided", () => {
		const { container } = render(
			<FriendsBlock
				state={{
					title: "Friends",
					count: 2,
					users: mockUsers,
				}}
				onClickBlock={() => {}}
			>
				<div>No friends</div>
			</FriendsBlock>,
		);
		expect(container).toMatchSnapshot();
		expect(screen.getByText("John")).toBeInTheDocument();
		expect(screen.getByText("Jane")).toBeInTheDocument();
	});

	it("renders children when no users are provided", () => {
		const { container } = render(
			<FriendsBlock
				state={{
					title: "Friends",
					count: 0,
					users: [],
				}}
				onClickBlock={() => {}}
			>
				<div>No friends</div>
			</FriendsBlock>,
		);
		expect(container).toMatchSnapshot();
		expect(screen.getByText("No friends")).toBeInTheDocument();
	});

	it("calls onClickBlock when title is clicked", () => {
		const handleClick = jest.fn();
		const { container } = render(
			<FriendsBlock
				state={{
					title: "Friends",
					count: 2,
					users: mockUsers,
				}}
				onClickBlock={handleClick}
			>
				<div>No friends</div>
			</FriendsBlock>,
		);
		expect(container).toMatchSnapshot();
		screen.getByText("Friends").click();
		expect(handleClick).toHaveBeenCalled();
	});
});
