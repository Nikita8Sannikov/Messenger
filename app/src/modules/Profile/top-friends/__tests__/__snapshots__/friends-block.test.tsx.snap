// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FriendsBlock calls onClickBlock when title is clicked 1`] = `
<div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row grid-item-style css-1fzlhpv-MuiGrid2-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 friends-container paper-block css-1gtchvp-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="block-title"
      >
        Friends
         
        <span
          class="counter"
        >
          2
        </span>
      </div>
      <ul
        class="MuiList-root MuiList-padding friends-container__top-friends__list css-b7dj7n-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding friends-container__top-friends__item css-cfd5wn-MuiListItem-root"
        >
          <div
            class="MuiListItemAvatar-root friends-container__top-friends__item__avatar-block css-1ulznza-MuiListItemAvatar-root"
          >
            <span
              class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular avatar user-avatar big-avatar css-fmeqll-MuiAvatar-root"
                id="user-avatar"
              >
                <img
                  alt="John Doe Smith"
                  class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                  src="avatar1.jpg"
                />
              </div>
              <span
                class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
              />
            </span>
          </div>
          John
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding friends-container__top-friends__item css-cfd5wn-MuiListItem-root"
        >
          <div
            class="MuiListItemAvatar-root friends-container__top-friends__item__avatar-block css-1ulznza-MuiListItemAvatar-root"
          >
            <span
              class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular avatar user-avatar big-avatar css-fmeqll-MuiAvatar-root"
                id="user-avatar"
              >
                <img
                  alt="Jane Smith Doe"
                  class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                  src="avatar2.jpg"
                />
              </div>
              <span
                class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
              />
            </span>
          </div>
          Jane
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`FriendsBlock renders children when no users are provided 1`] = `
<div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row grid-item-style css-1fzlhpv-MuiGrid2-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 friends-container paper-block css-1gtchvp-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="block-title"
      >
        Friends
         
        <span
          class="counter"
        >
          0
        </span>
      </div>
      <div>
        No friends
      </div>
    </div>
  </div>
</div>
`;

exports[`FriendsBlock renders loading state 1`] = `
<div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row grid-item-style css-1fzlhpv-MuiGrid2-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 friends-container paper-block css-1gtchvp-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="block-title"
      >
        Friends
         
        <span
          class="counter"
        >
          2
        </span>
      </div>
      <div
        class="spinner"
        data-testid="spinner"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`FriendsBlock renders users list when users are provided 1`] = `
<div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row grid-item-style css-1fzlhpv-MuiGrid2-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 friends-container paper-block css-1gtchvp-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="block-title"
      >
        Friends
         
        <span
          class="counter"
        >
          2
        </span>
      </div>
      <ul
        class="MuiList-root MuiList-padding friends-container__top-friends__list css-b7dj7n-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding friends-container__top-friends__item css-cfd5wn-MuiListItem-root"
        >
          <div
            class="MuiListItemAvatar-root friends-container__top-friends__item__avatar-block css-1ulznza-MuiListItemAvatar-root"
          >
            <span
              class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular avatar user-avatar big-avatar css-fmeqll-MuiAvatar-root"
                id="user-avatar"
              >
                <img
                  alt="John Doe Smith"
                  class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                  src="avatar1.jpg"
                />
              </div>
              <span
                class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
              />
            </span>
          </div>
          John
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding friends-container__top-friends__item css-cfd5wn-MuiListItem-root"
        >
          <div
            class="MuiListItemAvatar-root friends-container__top-friends__item__avatar-block css-1ulznza-MuiListItemAvatar-root"
          >
            <span
              class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular avatar user-avatar big-avatar css-fmeqll-MuiAvatar-root"
                id="user-avatar"
              >
                <img
                  alt="Jane Smith Doe"
                  class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                  src="avatar2.jpg"
                />
              </div>
              <span
                class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
              />
            </span>
          </div>
          Jane
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`FriendsBlock renders with title and count 1`] = `
<div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row grid-item-style css-1fzlhpv-MuiGrid2-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 friends-container paper-block css-1gtchvp-MuiPaper-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="block-title"
      >
        Friends
         
        <span
          class="counter"
        >
          2
        </span>
      </div>
      <ul
        class="MuiList-root MuiList-padding friends-container__top-friends__list css-b7dj7n-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding friends-container__top-friends__item css-cfd5wn-MuiListItem-root"
        >
          <div
            class="MuiListItemAvatar-root friends-container__top-friends__item__avatar-block css-1ulznza-MuiListItemAvatar-root"
          >
            <span
              class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular avatar user-avatar big-avatar css-fmeqll-MuiAvatar-root"
                id="user-avatar"
              >
                <img
                  alt="John Doe Smith"
                  class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                  src="avatar1.jpg"
                />
              </div>
              <span
                class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
              />
            </span>
          </div>
          John
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding friends-container__top-friends__item css-cfd5wn-MuiListItem-root"
        >
          <div
            class="MuiListItemAvatar-root friends-container__top-friends__item__avatar-block css-1ulznza-MuiListItemAvatar-root"
          >
            <span
              class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular avatar user-avatar big-avatar css-fmeqll-MuiAvatar-root"
                id="user-avatar"
              >
                <img
                  alt="Jane Smith Doe"
                  class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                  src="avatar2.jpg"
                />
              </div>
              <span
                class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
              />
            </span>
          </div>
          Jane
        </li>
      </ul>
    </div>
  </div>
</div>
`;
